staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf/collections.go

```diff
@@ -0,0 +1,202 @@
+	}
+	if listType.Field(0).Tag.Get("protobuf") != "" {
+		return metav1.TypeMeta{}, metav1.ListMeta{}, nil, fmt.Errorf(`expected TypeMeta json field tag to be ""`)
+	}
+	// ListMeta
+	listMeta, ok := listValue.Field(1).Interface().(metav1.ListMeta)
+	if !ok {
+		return metav1.TypeMeta{}, metav1.ListMeta{}, nil, fmt.Errorf("expected ListMeta field to have ListMeta type")
+	}
+	if listType.Field(1).Tag.Get("protobuf") != "bytes,1,opt,name=metadata" {
+		return metav1.TypeMeta{}, metav1.ListMeta{}, nil, fmt.Errorf(`expected ListMeta json field tag to be "bytes,1,opt,name=metadata"`)
+	}
+	// Items
+	items, err := meta.ExtractList(list)
+	if err != nil {
+		return metav1.TypeMeta{}, metav1.ListMeta{}, nil, err
+	}
+	if listType.Field(2).Tag.Get("protobuf") != "bytes,2,rep,name=items" {
+		return metav1.TypeMeta{}, metav1.ListMeta{}, nil, fmt.Errorf(`expected Items json field tag to be "bytes,2,rep,name=items"`)
+	}
+	return typeMeta, listMeta, items, nil
+}
+
+func streamingEncodeUnknownList(w io.Writer, unk runtime.Unknown, listMeta metav1.ListMeta, items []runtime.Object) error {
+	_, err := w.Write(protoEncodingPrefix)
+	if err != nil {
+		return err
+	}
+	// TypeMeta
+	_, err = encodeValue(w, []byte{0xa}, &unk.TypeMeta)
+	if err != nil {
+		return err
+	}
+	// Raw
+	_, err = w.Write([]byte{0x12})
+	if err != nil {
+		return err
+	}
+	err = streamingEncodeList(w, listMeta, items)
+	if err != nil {
+		return err
+	}
+	// ContentEncoding
+	_, err = encodeValue(w, []byte{0x1a}, unk.ContentEncoding)
+	if err != nil {
+		return err
+	}
+	// ContentType
+	_, err = encodeValue(w, []byte{0x22}, unk.ContentType)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func streamingEncodeList(w io.Writer, listMeta metav1.ListMeta, items []runtime.Object) error {
+	size := sizeWithHeader(&listMeta) + itemsSize(items)
+	// ListMeta
+	_, err := writeVarintGenerated(w, size)
+	if err != nil {
+		return err
+	}
+	listMetaSize, err := encodeValue(w, []byte{0xa}, &listMeta)
+	if err != nil {
+		return err
+	}
+	// Items
+	listSize, err := streamingEncodeObjectSlice(w, items)
+	if err != nil {
+		return err
+	}
+	if listMetaSize+listSize != size {
+		return fmt.Errorf("the size value was %d, but encoding wrote %d bytes to data", size, listSize)
+	}
+	return nil
+}
+
+func streamingEncodeObjectSlice(w io.Writer, items []runtime.Object) (size int, err error) {
+	for _, item := range items {
+		n, err := encodeValue(w, []byte{0x12}, item)
+		size += n
+		if err != nil {
+			return size, err
+		}
+	}
+	return size, nil
+}
+
+func encodeValue(w io.Writer, field []byte, value any) (size int, err error) {
+	size, err = w.Write(field)
+	if err != nil {
+		return size, err
+	}
+	switch v := value.(type) {
+	case proto.Marshaler:
+		data, err := v.Marshal()
+		if err != nil {
+			return size, err
+		}
+		n, err := writeVarintGenerated(w, len(data))
+		if err != nil {
+			return size, err
+		}
+		size += n
```

Comment:
Should we always `+n` before return error to keep consistent with other places?
```suggestion
		n, err := writeVarintGenerated(w, len(data))
		size += n
		if err != nil {
			return size, err
		}
```